pkgs/build-support/trivial-builders/default.nix

```diff
@@ -815,6 +815,36 @@ rec {
         writeDirectReferencesToFile (writeText "string-file" string);
 
 
+  /**
+   * Write the references (i.e. the runtime dependencies in the Nix store) of the elements in `paths' to a file.
+   */
+  writeMultipleReferencesToFile = paths:
+    let
+      # Dump the string content / output paths of all the member of `paths` into a file
+      pathListFile = writeText "path-list" ((builtins.concatStringsSep "\n"
+        # Uniquely sort to prevent unnessesary rebuilds
+        # Use string interpolation instead of `toString`
+        # to interpolate path values and path-like expressions.
+        (lib.unique (builtins.sort builtins.lessThan (map (p: "${p}") paths)))
+      ) + "\n");
+    in
+    # Get the referenecs of pathListFile, but
+    # remove the line containing the path to pathListFile
+    (writeReferencesToFile pathListFile).overrideAttrs (previousAttrs: {
```

Comment:
I've recently learned, that `exportReferencesGraph` offers a much cleaner interface when you use `__structuredAttrs = true`: https://github.com/NixOS/nixpkgs/blob/c2e4139f043eb5d1f0f679288ec965fe1c0688cf/pkgs/build-support/closure-info.nix#L16-L18 https://github.com/NixOS/nixpkgs/blob/c2e4139f043eb5d1f0f679288ec965fe1c0688cf/pkgs/build-support/closure-info.nix#L36-L38

Admittedly, this is a very surface-level comment, but I think we need to answer the following questions prior to merging:

1. Can we later rewrite `writeMultipleReferencesToFile` using the json interface for `exportReferencesGraph` while preserving the interface introduced in this PR?
2. Does `writeMultupleReferencesToFile` not duplicate e.g. `closureInfo`?

Reply:
Oh, I thought this was used to compute the closure for singularity images? Sorry, I'm completely forgetting the context, this has been open for so long...

Reply:
> 2\. Does `writeMultupleReferencesToFile` not duplicate e.g. `closureInfo`?

Not even close.
```
nixpkgs_timeshift (9b54fb4) [$]
❯ nix build --no-link --print-out-paths --impure --expr "let pkgs = import ./. { }; in pkgs.closureInfo { rootPaths = [ pkgs.linux ]; }"
/nix/store/sd8vsqzxhq9x4hyz3fvvggss0i4wkqvi-closure-info

nixpkgs_timeshift (9b54fb4) [$]
❯ ls -A /nix/store/sd8vsqzxhq9x4hyz3fvvggss0i4wkqvi-closure-info
registration  store-paths  total-nar-size

nixpkgs_timeshift (9b54fb4) [$]
❯ cat /nix/store/sd8vsqzxhq9x4hyz3fvvggss0i4wkqvi-closure-info/registration
/nix/store/lgkpyy1b1zzh04n8b6andg8iw8vnr07m-linux-6.6.21
sha256:0m59hwzpdbv4ihbgcbgd23a8gbn3gnzx2igsygm0a6wjyz8r0pb6
131908608

0

nixpkgs_timeshift (9b54fb4) [$]
❯ cat /nix/store/sd8vsqzxhq9x4hyz3fvvggss0i4wkqvi-closure-info/store-paths
/nix/store/lgkpyy1b1zzh04n8b6andg8iw8vnr07m-linux-6.6.21

nixpkgs_timeshift (9b54fb4) [$]
❯ cat /nix/store/sd8vsqzxhq9x4hyz3fvvggss0i4wkqvi-closure-info/total-nar-size
131908608

```

Reply:
> Can we later rewrite `writeMultipleReferencesToFile` using the json interface for `exportReferencesGraph` while preserving the interface introduced in this PR?

I hope so. I was originally plan to implement this function with `exportReferencesGraph`. It is the confusing yet undocumented format that makes me turn to `writeReferencesToFile`.

Reply:
Update: I finally got an idea how to read the files generated by `exportReferencesGraph`, but the implementation will become much closer to the one based on Bash associative array (dictionary).

Reply:
> Nothing is produced when passing __structuredAttrs = true;

Check `$NIX_ATTRS_JSON_FILE`/`.attrs.json`.

> with exportReferencesGraph = [ "graph" pkgs.hello ]

Looking at `closure-info.nix`, I guess this should be `exportReferencesGraph.graph = [ pkgs.hello ]` when using structured attrs

Reply:
> >    2. Does writeMultupleReferencesToFile not duplicate e.g. closureInfo?
> Not even close.

(I'm half asleep and not sure what I'm looking at but) maybe the right one was `pkgs/build-support/references-by-popularity/default.nix`...

Reply:
Looking at the occurrences of "`unnecessaryDrvs`" in `pkgs/build-support/docker/default.nix` I suspect they the same logic you've implemented in your `writeMultipleReferencesToFile` but scattered across `dockerTools` and `referencesByPopularity` (plus the sort thing)

Reply:
I'm not sure why would `referencesByPopularity` set `exportReferences.graph` to a `Path` instead of a `List Path`, which the other derivation did. If that accepts lists, there's no reason to do the "write a file with all the root paths, then filter that file out from the closure" trick they do in `dockerTools`' `default.nix`

Reply:
> Update: I finally got an idea how to read the files generated by exportReferencesGraph, but the implementation will become much closer to the one based on Bash associative array (dictionary).

Jq/python sound more promising. We don't have to rewrite now, but we do need to decide whether to introduce the new function or somehow extend the utils used by `dockerTools` (`referencesByPopularity`)

Reply:
> > Nothing is produced when passing __structuredAttrs = true;
> 
> Check `$NIX_ATTRS_JSON_FILE`/`.attrs.json`.
> 
> > with exportReferencesGraph = [ "graph" pkgs.hello ]
> 
> Looking at `closure-info.nix`, I guess this should be `exportReferencesGraph.graph = [ pkgs.hello ]` when using structured attrs

It's totally undocumented! Thanks for pointing out.

Reply:
@SomeoneSerge Thank you for telling me about the beautiful `exportReferencesGraph` interface with `__structuredAttrs` enabled. Now the implementation is as elegant as a Nix expression!